File: pystencil/pystencil/editor/edits.py
from __future__ import annotations

"""The chainable edits: rotation, crop, filter/tint and the x/y coordinate formulas.

Each mutator snapshots history through the ``_push`` that :class:`Editor` owns and
returns ``self``.
"""

from typing import Optional, Tuple

from ._snapshot import _Snapshot


class _EditApi:
  """Rotate / crop / filter / formula mutators."""

  # ── edits (chainable; each snapshots history) ──────────────────────────────
  def rotate(self, quarters: int) -> "Editor":
    """Rotate by ``quarters`` clockwise quarter-turns; the crop rect rides along."""
    self._require_original()
    core = self._get_core()
    cur = self._current()
    nxt = cur.copy()
    if cur.crop is not None:
      # Map the crop into the post-rotation rotated-original space (session.applyRotate).
      orig = self._original
      dims_w, dims_h = core.rotated_dims(orig.width, orig.height, cur.rotation)
      nxt.crop = self._rotate_rect_quarters(cur.crop, dims_w, dims_h, quarters)
    nxt.rotation = core.normalize_quarters(cur.rotation + quarters)
    self._push(nxt)
    return self

  def rotate_left(self) -> "Editor":
    """Rotate one quarter-turn counter-clockwise (-1)."""
    return self.rotate(-1)

  def rotate_right(self) -> "Editor":
    """Rotate one quarter-turn clockwise (+1)."""
    return self.rotate(1)

  def crop(
    self,
    spec: Optional[str] = None,
    *,
    x1: Optional[float] = None,
    y1: Optional[float] = None,
    x2: Optional[float] = None,
    y2: Optional[float] = None,
    album: bool = False,
  ) -> "Editor":
    """Crop the current view by a crop spec (or x1/y1/x2/y2 edges).

    When ``spec`` is omitted, a ``"x1=.. y1=.. x2=.. y2=.."`` spec is built from the
    given edges (None edges are omitted). The spec resolves against the CURRENT view's
    dimensions and page metrics (mirroring ``pipeline.resolveCropSpec`` →
    ``pageForImage``), then composes into rotated-original space exactly like
    ``session.applyCrop``. An unparseable spec is a no-op (matching the CLI, which
    prints an error and leaves the image unchanged).
    """
    self._require_original()
    core = self._get_core()
    cur = self._current()
    if spec is None:
      spec = self._build_crop_spec(x1, y1, x2, y2)
    rect = self.resolve_crop_rect(spec, album=album)
    if rect is None:
      # Bad spec: leave the editor untouched, just like the Zig handler.
      return self
    rx, ry, rw, rh = rect
    # The view is rotate(original) cropped to cur.crop; a sub-rect maps back by origin.
    base_x = cur.crop[0] if cur.crop is not None else 0
    base_y = cur.crop[1] if cur.crop is not None else 0
    orig = self._original
    space_w, space_h = core.rotated_dims(orig.width, orig.height, cur.rotation)
    new_crop = self._clamp_rect(
      (base_x + rx, base_y + ry, rw, rh), space_w, space_h
    )
    nxt = cur.copy()
    nxt.crop = new_crop
    self._push(nxt)
    return self

  def resolve_crop_rect(
    self, spec: str, *, album: bool = False
  ) -> Optional[Tuple[int, int, int, int]]:
    """Resolve a crop spec against the CURRENT view without applying it.

    Exactly the resolution :meth:`crop` performs — the same core ``resolveCrop``
    call against the same view dimensions and page metrics — returning the
    ``(x, y, w, h)`` sub-rect of the current view the crop would keep, or ``None``
    for an unparseable spec. The rect's origin is what the LLM executor's §1
    coordinate re-mapping subtracts from later plan coordinates.
    """
    self._require_original()
    core = self._get_core()
    view_w, view_h = self._view_dims(self._current())
    page_w, page_h = self._page_for_image(view_w, view_h)
    return core.resolve_crop(
      spec, view_w, view_h, view_w / page_w, view_h / page_h,
      page_w, page_h, album,
    )

  def set_filter(self, mode: str) -> "Editor":
    """Set the filter mode ("none"|"bw"|"sepia"|"custom"|"invert"|"contour"),
    keeping any custom colour."""
    self._require_original()
    cur = self._current()
    nxt = cur.copy()
    nxt.filter_mode = mode
    self._push(nxt)
    return self

  def set_filter_color(self, color: str) -> "Editor":
    """Set the custom duotone tint (hex); switches the mode to "custom"."""
    self._require_original()
    core = self._get_core()
    # Normalize through the core so "red"/"#f00"/"#ff0000" all land as a #rrggbb hex,
    # matching applyFilterArg's custom branch in the Zig handlers.
    parsed = core.parse_color(color)
    hex_color = "#%02x%02x%02x" % (parsed[0], parsed[1], parsed[2]) if parsed else color
    cur = self._current()
    nxt = cur.copy()
    nxt.filter_mode = "custom"
    nxt.filter_color = hex_color
    self._push(nxt)
    return self

  def apply_filter(self, mode: str) -> "Editor":
    """Convenience filter setter mirroring the CLI's ``/filter`` (``applyFilterArg``).

    "bw"/"sepia"/"invert"/"contour"/"none" set those modes directly (the named modes
    are checked BEFORE the colour fallback); anything else is treated as a colour —
    parsed by the core and stored as a custom #rrggbb duotone tint. An unrecognized
    value raises ``ValueError``.
    """
    low = mode.strip().lower()
    if low in ("bw", "sepia", "invert", "contour", "none"):
      return self.set_filter(low)
    core = self._get_core()
    parsed = core.parse_color(mode.strip())
    if parsed is None:
      raise ValueError(
        "unknown filter %r — use 'bw', 'sepia', 'invert', 'contour', 'none', "
        "or a colour" % mode
      )
    return self.set_filter_color(mode.strip())

  # ── formulas (x/y coordinate transform) ─────────────────────────────────────
  def set_formula(self, axis: str, expr: str) -> "Editor":
    """Set the x or y coordinate-transform formula (validated by the shared parser; raises
    ValueError on a bad expression). A non-empty formula enables formulas. The expression
    rides the saved layout, where the browser applies it; use apply_formula() to evaluate."""
    ax = "y" if axis == "y" else "x"
    expr = (expr or "").strip()
    if expr and not self._get_core().validate_formula(expr, ax):
      raise ValueError(f"invalid {ax} formula: {expr!r}")
    if ax == "y":
      self._formula_y = expr
    else:
      self._formula_x = expr
    if expr:
      self._allow_formulas = True
    return self

  def set_allow_formulas(self, on: bool) -> "Editor":
    """Toggle whether formulas apply (keeps the expressions, so re-enabling restores them)."""
    self._allow_formulas = bool(on)
    return self

  @property
  def allow_formulas(self) -> bool:
    """Whether the x/y formulas are currently applied."""
    return self._allow_formulas

  def apply_formula(self, axis: str, value: float) -> float:
    """Apply the current x or y formula to a coordinate (identity when off/empty/invalid)."""
    ax = "y" if axis == "y" else "x"
    expr = self._formula_y if ax == "y" else self._formula_x
    return self._get_core().apply_formula(expr, ax, value, self._allow_formulas)

File: cli/src/console/session/edits.zig
//! The structured edit state — crop, rotation, filter, drawn lines, page format and the
//! coordinate formulas — and the browser-compatible layout JSON it serializes to.
const Session = @import("../session.zig").Session;
const session_mod = @import("../session.zig");
const jsonStr = session_mod.jsonStr;
const jsonInt = session_mod.jsonInt;
const jsonNum = session_mod.jsonNum;
const std = @import("std");
const server = @import("../../server/client.zig");
const core = @import("../../core.zig");
const pipeline = @import("../../pipeline.zig");
const page_mod = @import("../../media/page.zig");
const EditState = @import("../session.zig").EditState;
const clampRect = @import("../session.zig").clampRect;
const rotateRectQuarters = @import("../session.zig").rotateRectQuarters;
const extractLinesJson = @import("../session.zig").extractLinesJson;
const mergeLinesJson = @import("../session.zig").mergeLinesJson;
const parseLayoutInto = @import("../session.zig").parseLayoutInto;

// editing ops (each pushes a snapshot + rebuilds)

/// Rotate by `n` quarter-turns (clockwise). The crop rect rides along into the new space.
pub fn applyRotate(self: *Session, n: i32) !void {
    const cur = self.state();
    var next = try cur.dupe(self.gpa);
    errdefer next.deinit(self.gpa);
    if (next.crop) |cr| {
        const orig = self.original.?;
        const dims = core.rotatedDims(@intCast(orig.width), @intCast(orig.height), cur.rotation);
        next.crop = rotateRectQuarters(cr, dims.w, dims.h, n);
    }
    next.rotation = core.normalizeQuarters(cur.rotation + n);
    try self.pushState(next);
}

/// Crop to `rect` (given in CURRENT-view pixels); composes into rotated-original space.
pub fn applyCrop(self: *Session, rect: core.Rect) !void {
    const cur = self.state();
    var next = try cur.dupe(self.gpa);
    errdefer next.deinit(self.gpa);
    // The view is rotate(original) cropped to `cur.crop`; a sub-rect maps back by its origin.
    const base_x: i32 = if (cur.crop) |c| c.x else 0;
    const base_y: i32 = if (cur.crop) |c| c.y else 0;
    const orig = self.original.?;
    const dims = core.rotatedDims(@intCast(orig.width), @intCast(orig.height), cur.rotation);
    next.crop = clampRect(.{ .x = base_x + rect.x, .y = base_y + rect.y, .w = rect.w, .h = rect.h }, dims.w, dims.h);
    try self.pushState(next);
}

/// Set the image filter (mode "none"|"bw"|"sepia"|"custom"; color is the custom hex).
pub fn setFilter(self: *Session, mode: []const u8, color: []const u8) !void {
    const cur = self.state();
    var next = try cur.dupe(self.gpa);
    errdefer next.deinit(self.gpa);
    if (next.filter_mode.len != 0) self.gpa.free(next.filter_mode);
    if (next.filter_color.len != 0) self.gpa.free(next.filter_color);
    next.filter_mode = try self.gpa.dupe(u8, mode);
    next.filter_color = try self.gpa.dupe(u8, color);
    try self.pushState(next);
}

/// Append the lines from a layout JSON document to the drawing.
pub fn addLines(self: *Session, layout_bytes: []const u8) !void {
    const add = try extractLinesJson(self.gpa, layout_bytes);
    defer self.gpa.free(add);
    const cur = self.state();
    var next = try cur.dupe(self.gpa);
    errdefer next.deinit(self.gpa);
    const merged = try mergeLinesJson(self.gpa, cur.lines(), add);
    if (next.lines_json.len != 0) self.gpa.free(next.lines_json);
    next.lines_json = merged;
    try self.pushState(next);
}

/// Replace the drawing's lines with those from a layout JSON document — the
/// counterpart to `addLines` (which appends), for `apply <src> replace`.
pub fn setLines(self: *Session, layout_bytes: []const u8) !void {
    const add = try extractLinesJson(self.gpa, layout_bytes);
    defer self.gpa.free(add);
    const cur = self.state();
    var next = try cur.dupe(self.gpa);
    errdefer next.deinit(self.gpa);
    const only = try self.gpa.dupe(u8, add);
    if (next.lines_json.len != 0) self.gpa.free(next.lines_json);
    next.lines_json = only;
    try self.pushState(next);
}

/// Adopt a server project's stored layout into the pristine state (used right after a
/// fetch/pull loads the original), so the view shows the peer's crop/rotation/filter/lines.
pub fn adoptServerLayout(self: *Session, layout_bytes: []const u8) !void {
    var st = EditState{};
    parseLayoutInto(self.gpa, layout_bytes, &st) catch {}; // partial parse still yields a valid st
    // Replace history[0] (we are right after loadImage, so cursor == 0). st is moved in.
    self.history.items[0].deinit(self.gpa);
    self.history.items[0] = st;
    self.cursor = 0;
    self.adoptLayoutMeta(layout_bytes); // page format + formulas (project-level, round-tripped)
    self.rebuild() catch {};
}

/// Parse the page format + x/y formulas out of a fetched layout (best-effort; cleared on miss).
pub fn adoptLayoutMeta(self: *Session, layout_bytes: []const u8) void {
    self.clearFormat();
    var parsed = std.json.parseFromSlice(std.json.Value, self.gpa, layout_bytes, .{}) catch return;
    defer parsed.deinit();
    if (parsed.value != .object) return;
    const obj = parsed.value.object;
    if (jsonStr(obj, "pageSize")) |ps| self.page_size = self.gpa.dupe(u8, ps) catch &.{};
    self.custom_page_w = jsonNum(obj, "customPageWidth");
    self.custom_page_h = jsonNum(obj, "customPageHeight");
    if (obj.get("allowFormulas")) |v| {
        if (v == .bool) self.allow_formulas = v.bool;
    }
    if (jsonStr(obj, "formulaX")) |fx| self.formula_x = self.gpa.dupe(u8, fx) catch &.{};
    if (jsonStr(obj, "formulaY")) |fy| self.formula_y = self.gpa.dupe(u8, fy) catch &.{};
}

/// The project-level page format + formulas as a PageMeta view (borrows the owned slices).
pub fn pageMeta(self: *Session) server.PageMeta {
    return .{
        .page_size = self.page_size,
        .custom_w = self.custom_page_w,
        .custom_h = self.custom_page_h,
        .allow_formulas = self.allow_formulas,
        .formula_x = self.formula_x,
        .formula_y = self.formula_y,
    };
}

/// Build the browser-compatible layout JSON for the current state (caller owns it).
pub fn currentLayoutJson(self: *Session) ![]u8 {
    const st = self.state();
    const img = self.current();
    // No explicit crop means the WHOLE rotated original — stated rather than omitted: the GUIs auto-crop
    // to the page aspect unless the layout names a cropRect, stranding lines drawn outside the page.
    const crop: ?server.CropRect = if (st.crop) |c|
        .{ .x = c.x, .y = c.y, .w = c.w, .h = c.h }
    else
        .{ .x = 0, .y = 0, .w = @intCast(img.width), .h = @intCast(img.height) };
    return server.buildLayout(self.gpa, @intCast(img.width), @intCast(img.height), st.lines(), st.filter_mode, st.filter_color, crop, st.rotation, self.pageMeta());
}

/// Page-format label beside the px size, e.g. "A4 21×29.7cm" (or "custom <w>×<h>cm"). Shares the one
/// derivation with the one-shot pipeline's wrote line (page.pageLabelAlloc); caller owns it.
pub fn pageFormatLabel(self: *Session) ![]u8 {
    const img = self.current();
    return pipeline.pageLabelAlloc(self.gpa, self.page_size, self.custom_page_w, self.custom_page_h, img.width, img.height);
}

/// Set the picked page format (canonical "A0".."C10" or "custom"), owned copy. It drives
/// the header label, the layout `pageSize` written on save/sync, and the /blank default.
pub fn setPageSize(self: *Session, name: []const u8) !void {
    const dup = try self.gpa.dupe(u8, name);
    if (self.page_size.len != 0) self.gpa.free(self.page_size);
    self.page_size = dup;
}

/// Replace the displayed label (e.g. after a rename), owned copy.
pub fn setLabel(self: *Session, name: []const u8) !void {
    const dup = try self.gpa.dupe(u8, name);
    if (self.label) |l| self.gpa.free(l);
    self.label = dup;
}

/// What a formula's names resolve to in this session: the page in cm and, once an image is
/// open, its pixels. The console has no display-unit switch, so PAGE_WIDTH reads cm.
pub fn formulaContext(self: *Session) core.FormulaCtx {
    const img = self.working;
    const w: usize = if (img) |i| i.width else 0;
    const h: usize = if (img) |i| i.height else 0;
    const page = page_mod.pageCmFor(self.page_size, self.custom_page_w, self.custom_page_h, w, h);
    var ctx = core.FormulaCtx{ .page_w_cm = page.w, .page_h_cm = page.h };
    if (img != null) {
        ctx.image_w = @floatFromInt(w);
        ctx.image_h = @floatFromInt(h);
    }
    return ctx;
}

/// Set the x or y transform formula (validated via the shared parser with this session's named
/// values in reach). Returns false on an invalid expression, state unchanged.
pub fn setFormula(self: *Session, axis: u8, expr: []const u8) !bool {
    // The context is built BEFORE the expression is copied: it resolves a page name through
    // core.zstr's one scratch buffer, which would otherwise overwrite the copy of `expr`.
    const ctx = formulaContext(self);
    if (expr.len != 0 and !core.validateFormulaCtx(core.zstr(expr) orelse return false, ctx)) return false;
    const dup = try self.gpa.dupe(u8, expr);
    const slot = if (axis == 'y') &self.formula_y else &self.formula_x;
    if (slot.len != 0) self.gpa.free(slot.*);
    slot.* = dup;
    if (expr.len != 0) self.allow_formulas = true;
    return true;
}

/// Toggle whether formulas apply on the saved layout (keeps the expressions).
pub fn setAllowFormulas(self: *Session, on: bool) void {
    self.allow_formulas = on;
}

/// Clear both formula expressions and disable formulas (keeps the page format).
pub fn clearFormulas(self: *Session) void {
    if (self.formula_x.len != 0) self.gpa.free(self.formula_x);
    if (self.formula_y.len != 0) self.gpa.free(self.formula_y);
    self.formula_x = &.{};
    self.formula_y = &.{};
    self.allow_formulas = false;
}

/// Reset the page format + formulas to "unset" (frees owned strings).
pub fn clearFormat(self: *Session) void {
    self.clearFormulas();
    if (self.page_size.len != 0) self.gpa.free(self.page_size);
    self.page_size = &.{};
    self.custom_page_w = 0;
    self.custom_page_h = 0;
}
